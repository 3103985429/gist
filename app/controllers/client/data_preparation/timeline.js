const winston = require('winston');
const { cond, constant, get, stubTrue } = require('lodash');

const { VIEW_TYPE_TIMELINE } = require('../../../config/views.config');
const { calculatePageRanges } = require('./data_prep_helpers/pagination_helpers');
const { Lazy_Shared_ProcessedRowObject_MongooseContext } = require('../../../models/processed_row_objects');
const { getTimelineSort } = require('./data_prep_helpers/sort.helpers');
const { getTimelineProject } = require('./data_prep_helpers/project.helpers');
const {
    getGroupByDurationParams,
    getGroupByRangeParams,
    getGroupByColumnParams,
} = require('./data_prep_helpers/group.helpers');
const { processDocs } = require('./data_prep_helpers/gallery.helpers');
const { processGroupedResults, getCountAggregationOperators } = require('./data_prep_helpers/timeline.helpers');
const {
    addDefaultAggregationOperators,
    aggregateProcessedRowObjects,
    getFieldTypeAggregationOperators,
} = require('./data_prep_helpers/aggregation.helpers');
const func = require('../func');

module.exports = { bindData };

async function bindData(dataSource, options) {
    const {
        filterObj,
        groupBy_realColumnName,
        groupSize,
        limit,
        onPageNum,
        pageNumber,
        routePath_base,
        sortBy_realColumnName,
        groupBy_isDate,
        groupBy_isNumber,
        groupByRange,
        groupByDuration,
    } = options;

    const timelineViewSettings = dataSource.fe_views.views.timeline;
    const { Model: processedRowObjects_mongooseModel } = Lazy_Shared_ProcessedRowObject_MongooseContext(dataSource._id);
    const groupByRealColumnNamePath = `rowParams.${groupBy_realColumnName}`;
    const wholeFilteredSetAggregationOperators = addDefaultAggregationOperators(dataSource, options);
    const fieldCheckMatch = getFieldTypeAggregationOperators(groupByRealColumnNamePath, {
        isDate: groupBy_isDate,
        isNumber: groupBy_isNumber,
    });

    wholeFilteredSetAggregationOperators.unshift(fieldCheckMatch);
    const getGroupParams = cond([
        [constant(groupBy_isDate), () => getGroupByDurationParams(groupBy_realColumnName, groupByDuration)],
        [constant(groupBy_isNumber), () => getGroupByRangeParams(groupBy_realColumnName, groupByRange)],
        [stubTrue, () => getGroupByColumnParams(groupBy_realColumnName)],
    ]);

    const groupParams = getGroupParams();
    const { groupIdQueryHelper, groupByDateFormat, groupIdProject } = groupParams;

    winston.debug('Getting nonPagedCount');

    const countWholeFilteredSetAggregationOperators = getCountAggregationOperators(
        wholeFilteredSetAggregationOperators,
        groupParams,
        options
    );

    const nonPagedResults = await aggregateProcessedRowObjects(
        processedRowObjects_mongooseModel,
        countWholeFilteredSetAggregationOperators
    );
    const nonPagedCount = groupSize === -1 ? get(nonPagedResults, [0, 'count'], 0) : nonPagedResults.length;

    winston.debug('Found %i non-paged docs', nonPagedCount);

    const skipNResults = limit * (Math.max(pageNumber, 1) - 1);

    const sort = getTimelineSort(dataSource, options);
    const additionalProject = getTimelineProject(dataSource, options);
    const project = {
        _id: 1,
        pKey: 1,
        ...additionalProject,
    };
    const resultsProject = { $slice: groupSize === -1 ? ['$results', skipNResults, limit] : ['$results', groupSize] };

    const getAggregationGroupParams = cond([
        [
            constant(groupBy_isDate),
            constant({
                startDate: { $min: `$${groupByRealColumnNamePath}` },
                endDate: { $max: `$${groupByRealColumnNamePath}` },
            }),
        ],
        [constant(groupBy_isNumber), constant({ group: { $first: `$${groupByRealColumnNamePath}` } })],
        [stubTrue, constant({ group: { $first: `$${groupByRealColumnNamePath}` } })],
    ]);
    const getAggregationProjectParams = cond([
        [
            constant(groupBy_isDate),
            constant({
                startDate: 1,
                endDate: 1,
            }),
        ],
        [constant(groupBy_isNumber), constant({ group: groupIdProject })],
        [stubTrue, constant({ group: 1 })],
    ]);
    const sortGroupParams = groupBy_isDate ? { startDate: -sort._id } : { group: sort._id };

    const aggregationGroupParams = getAggregationGroupParams();
    const aggregationProjectParams = getAggregationProjectParams();
    const unwind = [{ $unwind: `$${groupByRealColumnNamePath}` }];

    if (!groupBy_isDate && !groupBy_isNumber) {
        unwind.push({ $unwind: `$${groupByRealColumnNamePath}` });
    }

    const aggregationOperators = wholeFilteredSetAggregationOperators.concat([
        { $project: project },
        ...unwind,
        { $sort: sort },
        {
            $group: {
                _id: groupIdQueryHelper,
                total: { $sum: 1 },
                results: {
                    $push: {
                        _id: '$_id',
                        rowParams: '$rowParams',
                        pKey: '$pKey',
                    },
                },
                ...aggregationGroupParams,
            },
        },
        {
            $project: {
                _id: 0,
                total: 1,
                results: resultsProject,
                ...aggregationProjectParams,
            },
        },
        { $sort: sortGroupParams },
    ]);

    if (groupSize !== -1) {
        aggregationOperators.push({ $skip: skipNResults });
        aggregationOperators.push({ $limit: limit });
    }

    const docs = await aggregateProcessedRowObjects(processedRowObjects_mongooseModel, aggregationOperators);
    const groupedResults = docs.map(group => {
        group.results = processDocs(group.results, dataSource, VIEW_TYPE_TIMELINE, timelineViewSettings, options);

        return group;
    });
    const processedResults = processGroupedResults(groupedResults, dataSource, options, groupParams);
    const pageRanges = calculatePageRanges(nonPagedCount);

    const isPercentValue = columnName => func.isPercentOperation(dataSource, columnName);
    const isGroupByPercent = isPercentValue(groupBy_realColumnName);
    const isSortByPercent = isPercentValue(sortBy_realColumnName);

    const data = {
        data: processedResults,
        included: {
            dataSource,
        },
        meta: {
            numberOfResults: groupedResults.length,
            numPages: Math.ceil(nonPagedCount / limit),
            nonpagedCount: nonPagedCount,
            pageRanges,
            onPageNum,
            routePath_base,
            filterObj,
            groupByDateFormat,
            groupSize,
            hasImages: true,
            isGroupByPercent,
            isSortByPercent,
        },
    };

    winston.debug('Done preparing data for timeline');

    return data;
}
