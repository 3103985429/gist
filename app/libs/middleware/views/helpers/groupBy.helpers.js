const { cond, constant, eq, has, stubTrue } = require('lodash');

const { RealColumnNameFromHumanReadableColumnName } = require('../../../datasources/imported_data_preparation');
const {
    SETTING_DEFAULT_GROUP_BY_DURATION,
    SETTING_DEFAULT_GROUP_BY_RANGE,
    SETTING_DEFAULT_GROUP_BY_COLUMN_NAME,
    SETTING_DEPENDS_ON_SORT_BY,
} = require('../../../../config/views.config');
const { fieldOverrideIfExists } = require('../../../../controllers/client/func');
const { equals } = require('../../../../utils/helpers');

const TYPE_DATE = 'date';
const TYPE_NUMBER = 'number';
const TYPE_STRING = 'string';

module.exports = {
    getDefaultGroupByColumnName,
    getGroupByTypeFromSortBy,
    invalidateGroupBy,
};

/**
 * @param {String} groupByType
 * @param {Object} dataSourceDescription
 * @param {String?} defaultGroupByDuration
 * @param {String?} defaultGroupByRange
 * @param {String?} defaultGroupByColumnName
 * @returns {String|undefined}
 */
function getDefaultGroupByColumnName(
    groupByType,
    dataSourceDescription,
    { defaultGroupByDuration, defaultGroupByRange, defaultGroupByColumnName }
) {
    if (groupByType === TYPE_DATE) {
        return defaultGroupByDuration;
    }

    if (groupByType === TYPE_NUMBER) {
        return defaultGroupByRange;
    }

    return fieldOverrideIfExists(defaultGroupByColumnName, dataSourceDescription);
}

/**
 * @param {Boolean} isDate
 * @param {Boolean} isNumber
 * @param {Object[]} masterViewSettings
 * @returns {String|null}
 */
function getGroupByTypeFromSortBy({ isDate, isNumber } = {}, masterViewSettings) {
    const findSetting = settingName => masterViewSettings.find(({ name }) => name === settingName);
    const getType = cond([
        [() => isDate && findSetting(SETTING_DEFAULT_GROUP_BY_DURATION), constant(TYPE_DATE)],
        [() => isNumber && findSetting(SETTING_DEFAULT_GROUP_BY_RANGE), constant(TYPE_NUMBER)],
        [
            () => has(findSetting(SETTING_DEFAULT_GROUP_BY_COLUMN_NAME), SETTING_DEPENDS_ON_SORT_BY),
            constant(TYPE_STRING),
        ],
        [stubTrue, constant(null)],
    ]);

    return getType();
}

function invalidateGroupBy(
    groupBy,
    groupByType,
    dataSourceDescription,
    { fieldsNotAvailableAsGroupByColumns = [], durationsAvailableForGroupBy = [], rangesAvailableForGroupBy = [] }
) {
    const { fe_excludeFields } = dataSourceDescription;
    const groupByRealColumnName = RealColumnNameFromHumanReadableColumnName(groupBy, dataSourceDescription, true);
    const stringGroupBy = () =>
        eq(fe_excludeFields[groupByRealColumnName], false) && !fieldsNotAvailableAsGroupByColumns[groupByRealColumnName]
            ? groupBy
            : undefined;
    const dateGroupBy = () => (durationsAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const numberGroupBy = () => (rangesAvailableForGroupBy.includes(groupBy) ? groupBy : undefined);
    const getDefaultGroupBy = cond([
        [equals(TYPE_STRING), stringGroupBy],
        [equals(TYPE_DATE), dateGroupBy],
        [equals(TYPE_NUMBER), numberGroupBy],
        [stubTrue, constant(groupBy)],
    ]);

    return getDefaultGroupBy(groupByType);
}
