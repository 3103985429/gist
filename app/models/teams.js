const integerValidator = require('mongoose-integer');
const winston = require('winston');
const { isEmpty } = require('lodash');

const mongoose_client = require('./mongoose_client');

const mongoose = mongoose_client.mongoose;
const Schema = mongoose.Schema;

const team_scheme = Schema({
    title: String,
    subdomain: { type: String, unique: true },
    description: String,
    font: { type: String, default: 'centrano2' },
    logo: String,
    logo_header: String,
    externalWebsite: String,
    colorPalette: {
        type: Array,
        default: [
            '#1B6DFC',
            '#0CB7FA',
            '#FA79FC',
            '#FB3705',
            '#F9D307',
            '#FA9007',
            '#7D5807',
            '#B2E606',
            '#069908',
            '#7B07FD',
        ],
    },
    superTeam: Boolean,
    admin: [{ type: Schema.Types.ObjectId, ref: 'User' }],
    datasourceDescriptions: [{ type: Schema.Types.ObjectId, ref: 'DatasourceDescription' }],
    pages: [{ type: Schema.Types.ObjectId, ref: 'Page' }],
    sites: [{ type: Schema.Types.ObjectId, ref: 'Website' }],

    isEnterprise: Boolean, //if true, means it has custom view - TODO consolidate with hasEnterpriseLicense
    customViewModal: Boolean, //if true, custom view has custom view settings modal - (rhodiumgroup)
    hasEnterpriseLicense: { type: Boolean, default: false },

    google_analytics: String,
    header_script: String,
    footer_script: String,
    brandColor: {
        label: { type: String, default: '#2A2A2A' },
        select: { type: String, default: '#2A2A2A' },
    },

}, { timestamps: true });

const modelName = 'Team';

team_scheme.plugin(integerValidator);

const deepPopulate = require('mongoose-deep-populate')(mongoose);

team_scheme.plugin(deepPopulate, { whitelist: ['datasourceDescriptions.author', 'datasourceDescriptions.updatedBy', 'datasourceDescriptions.createdAt'] });

team_scheme.pre('save', function (next) {
    this._wasNew = this.isNew;
    next();
});

team_scheme.methods.notifyNewTeamCreation = function () {
    if (this._wasNew) {
        this.populate('admin', function (err, docPopulatedWithAdmin) {
            if (err || isEmpty(docPopulatedWithAdmin.admin)) {
                winston.error('Team created with error err:' + err);
            }
        });
    }
};

team_scheme.statics.getTeamEntries = async (subdomain, user, vizName) => {
    const baseQuery = { subdomain };

    if (process.env.NODE_ENV === 'enterprise' && !process.env.SUBTEAMS) {
        baseQuery.subdomain = process.env.SUBDOMAIN;
    }

    const populateQuery = [
        {
            path: 'datasourceDescriptions',
            match: require('./descriptions').buildTeamPageQuery(user, vizName),
            select: 'description uid urls title importRevision updatedAt createdAt previous_datasets updatedBy author brandColor fe_views fe_filters.default banner connection',
            populate: [
                {
                    path: 'updatedBy',
                    select: 'firstName lastName',
                },
                {
                    path: 'author',
                    select: 'firstName lastName',
                },
                {
                    path: 'previous_datasets',
                    select: 'createdAt -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
        {
            path: 'pages',
            match: require('./pages').buildTeamPageQuery(user),
            select: 'pageTitle slug parsedSummary url thumbnail image updatedBy createdBy createdAt updatedAt createdAt',
            populate: [
                {
                    path: 'createdBy',
                    select: 'firstName lastName -_id',
                },
                {
                    path: 'website',
                    select: 'slug -_id',
                },
            ],
            options: {
                sort: {
                    createdAt: -1,
                },
            },
        },
    ];

    return team.find(baseQuery).populate(populateQuery);
};

const team = mongoose.model(modelName, team_scheme);

team.GetTeams = function (fn) {
    mongoose_client.WhenMongoDBConnected(function () {
        team.find({}, function (err, teams) {
            if (err) {
                fn(err);
            } else {
                fn(null, teams);
            }
        });
    });
};

module.exports = team;
