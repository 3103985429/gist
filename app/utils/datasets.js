const { findDataset } = require('../controllers/api/dataset/get-helpers');
const { some, includes } = require('lodash');
const DATA_SOURCE_TYPES = require('../../config/data-source-types');

module.exports.getDataset = getDataset;
module.exports.hasValidDatasetSource = hasValidDatasetSource;

/**
 * @param {String|ObjectId} id
 * @returns {Promise<Object>}
 */
async function getDataset(id) {
    const { description } = await findDataset({ id });

    return description;
}

/**
 * Determins if the datasourceDescription passed has a valid dataset source
 * determined from the list in config/data-source-types.js
 * @param {Object} datasourceDescription
 * @returns {Boolean}
 */
function hasValidDatasetSource(datasourceDescription) {
    return some(datasourceDescription, (value, key) => {
        return includes(DATA_SOURCE_TYPES, key);
    });
}
