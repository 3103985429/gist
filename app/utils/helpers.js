const { isEqual, trim, toString } = require('lodash');

module.exports.equals = a => b => isEqual(a, b);

module.exports.isEmbed = (query) => query.embed === 'true' || query.embed === true;

module.exports.formatColumnName = (value) => trim(toString(value).replace(/\./g, '_'));

module.exports.formatRawValues = (value) => trim(toString(value).replace(/’/g, '\'').replace(/”/g, '"'));

module.exports.addSlugIndex = (slugs, item, tilteProp = 'title') => {
    if (!slugs.length) {
        return;
    }

    let i = 2;
    const slugList = slugs.map((item) => item.slug);
    const max = i + slugList.length + 1;

    for (i; i < max; i++) {
        const newSlug = `${item.slug}-${i}`;

        if (!slugList.includes(newSlug)) {
            item.slug = newSlug;
            item[tilteProp] = `${item[tilteProp]} ${i}`;

            break;
        }
    }
};
