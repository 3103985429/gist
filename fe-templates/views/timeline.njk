{# TODO: why set all these variables like this? easier initial transition from the existing gallery templates.
    lots of opportunity to precalculate this stuff in the data_prep controller and slim down the templates. #}

{% set dataSource = included.dataSource %}
{% set viewOptions = dataSource.fe_viewOptions %}
{% set array_source_key = dataSource.uid %}
{% set fieldKey_objectTitle = options.fieldKey_objectTitle %}
{% set galleryStyle = options.galleryStyle %}
{% set secondaryColumnTileDisplay = options.secondaryColumnTileDisplay %}
{% set nonpagedCount = meta.nonpagedCount %}
{% set scrapedImages = meta.imageMeta.scrapedImages %}
{% set sharedPage = options.sharedPage %}
{% set routePath_base = options.routePath_base %}
{% set embedded = options.embedded %}
{% set isPreview = options.preview %}
{% set revision = options.revision %}
{% set hasThumbs = meta.imageMeta.hasThumbs %}
{% set timelineViewSettings = dataSource.fe_views.views.timeline %}
{% set displaySortFieldText = timelineViewSettings.displaySortFieldText %}
{% set displayFieldTitles = timelineViewSettings.displayFieldTitles %}
{% set team = dataSource._team if dataSource._team else null %}
{% set subdomain = team.subdomain | lower %}
{% set datasetId = dataSource._id %}
{% set viewType = 'timeline' %}
{% set sortBy = options.sortBy %}
{% set sortBy_realColumnName = options.sortBy_realColumnName %}
{% set groupBy_realColumnName = options.groupBy_realColumnName %}
{% set defaultSortByColumnName_humanReadable = options.defaultSortByColumnName_humanReadable %}
{% set secondaryCol = options.secondaryCol %}
{% set filterObj = options.filterObj %}
{% set imageCaption = options.largeTilesWithCaptions %}
{% set hasTooltip = not options.largeTilesWithCaptions %}
{% set groupBy_isDate = options.groupBy_isDate %}

<ul class="gallery-timeline">
    {% for group in data %}
        {% if groupBy_isDate %}
            {% set sectionTitle = group.startDate | dateFormat(meta.groupByDateFormat) %}

            {% if options.groupBy === 'Week' %}
                {% set sectionTitle = 'Week of ' + sectionTitle %}
            {% endif %}
        {% else %}
            {% set sectionTitle = group.group %}
        {% endif %}

        <li class="gallery-group {{ 'gallery-group--large' if options.largeTilesWithCaptions else 'gallery-group--small' }}"
            role="listbox" aria-label="{{ sectionTitle }}">
            <h3 class="gallery-group__label">{{ sectionTitle if sectionTitle !== '' else '(Title not specified)' }}</h3>

            <ul class="gallery-grid">
                {% for doc in group.results %}
                    {% set textField = sortBy_realColumnName %}

                    {% if textField === fieldKey_objectTitle %}
                        {% set displaySortFieldText = false %}
                    {% endif %}

                    {% if groupBy_isDate %}
                        {% set defaultTooltipValue = '(Date not specified)' %}
                    {% else %}
                        {% set defaultTooltipValue = '(Value not specified)' %}
                    {% endif %}

                    {% if hasTooltip %}
                        {% set tooltipKey = doc.rowParams[fieldKey_objectTitle] if doc.rowParams[fieldKey_objectTitle] else '(Title not specified)' %}
                        {% set tooltipValue = doc.rowParams[textField] if doc.rowParams[textField] else defaultTooltipValue %}
                    {% endif %}

                    {% if galleryStyle === 'image' and hasThumbs and doc.nestedImagePath %}
                        {% include 'views/partials/gallery_item_image.njk' %}
                    {% else %}
                        {% include 'views/partials/gallery_item_text.njk' %}
                    {% endif %}
                {% endfor %}

                {% if meta.groupSize < group.total and meta.groupSize >= 0 %}
                    <li class="gallery-item">
                        {% if options.viewOptions.viewInteractivity %}
                            <a href="{{ group.viewAllUrl }}" class="gallery-item-link gallery-item-link--view-all gallery-item-link--initial">
                                <div class="gallery-item-inner">
                                    <div class="gallery-item-initial-wrapper" aria-label="View All {{ group.total }}">
                                        <span class="gallery-item-initial gallery-item-initial--default">+{{ group.total }}</span>
                                        <span class="gallery-item-initial gallery-item-initial--hover">view all</span>
                                    </div>
                                </div>
                            </a>
                        {% else %}
                            <a class="gallery-item-link gallery-item-link--view-all gallery-item-link--initial">
                                <div class="gallery-item-inner">
                                    <div class="gallery-item-initial-wrapper" aria-label="{{ group.total }} more">
                                        <span class="gallery-item-initial gallery-item-initial--default">+{{ group.total }}</span>
                                        <span class="gallery-item-initial gallery-item-initial--hover">more</span>
                                    </div>
                                </div>
                            </a>
                        {% endif %}
                    </li>
                {% endif %}
            </ul>
        </li>
    {% endfor %}
</ul>
