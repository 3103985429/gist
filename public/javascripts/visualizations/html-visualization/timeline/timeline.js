/* global arrays, jQuery, nunjucks, d3 */

/**
 * External namespace for arrays classes
 * @external arrays
 */

/**
 * Creates a new Timeline
 * @constructor
 * @extends HTMLVisualization
 */
arrays.Timeline = function() {
    // Call the parent constructor, making sure (using Function#call)
    // that "this" is set correctly during the call
    arrays.HTMLVisualization.call(this);

    this._templateName = 'views/timeline.njk';
    this._meta = {};
    this._included = {};
    this._tooltip = new arrays.StandardTooltip();
};

arrays.Timeline.prototype = Object.create(arrays.HTMLVisualization.prototype);
arrays.Timeline.prototype.constructor = arrays.Timeline;

/**
 * Initialize the Timeline
 * @param {Object} data - All data required to render this visualization
 * @param {Object} options
 * @public
 * @extends HTMLVisualization
 * @returns {Timeline}
 */
arrays.Timeline.prototype.init = function(data, options) {
    arrays.HTMLVisualization.prototype.init.call(this, data, options);
    this.setMeta(data.meta);
    this.setIncluded(data.included);

    return this;
};

/**
 * Creates the static elements
 * @private
 * @extends Timeline
 */
arrays.Timeline.prototype._createStaticElements = function() {
    arrays.HTMLVisualization.prototype._createStaticElements.call(this);
    this._tooltip.setPosition('top');
};

/**
 * Renders a Timeline
 * @private
 */
arrays.Timeline.prototype._render = function() {
    const html = nunjucks.render(this._templateName, {
        data: this._data,
        meta: this._meta,
        included: this._included,
        options: this._options,
    });

    this._wrapper.append(html);

    const self = this;
    this._wrapper
        .find('.gallery-item.has-tooltip')
        .hover(
            function() {
                self.showTooltip(this);
            },
            function() {
                self.hideTooltip(this);
            }
        )
        .mousemove(function() {
            const tooltipDimensions = self._tooltip.getDimensions();
            self.moveTooltip(event.pageX - tooltipDimensions.width * 0.5, event.pageY - tooltipDimensions.height - 15);
        });
    this._wrapper
        .find('.gallery-item.has-tooltip a')
        .on('focus', function() {
            self.showTooltip(this);
        })
        .on('focusout', function() {
            self.hideTooltip(this);
        });
};

/**
 * Show tooltip.
 * @public
 */
arrays.Timeline.prototype.showTooltip = function(element) {
    const selection = $(element);
    this._hoveredElement = element;

    const tooltipValue =
        selection.data('tooltip-value') || selection.closest('.gallery-item.has-tooltip').data('tooltip-value');
    const value = _.isNil(tooltipValue) ? selection.text() : tooltipValue;
    const title = selection.data('tooltip-key') || selection.closest('.gallery-item.has-tooltip').data('tooltip-key');
    const templateOptions = {
        title: [arrays.escape(title)],
        rows: [[arrays.escape(value)]],
        arrow: true,
    };

    this._tooltip.applyTemplate(templateOptions).show(selection.closest('.gallery-item.has-tooltip')[0]);

    const centerX = selection.offset().left + selection.width() * 0.5;
    const centerY = selection.offset().top + selection.height() * 0.5;
    const tooltipDimensions = this._tooltip.getDimensions();
    this.moveTooltip(centerX - tooltipDimensions.width * 0.5, centerY - tooltipDimensions.height - 9);
};

/**
 * Hide tooltip.
 * @public
 */
arrays.Timeline.prototype.hideTooltip = function(element) {
    // prevent hide events from other elements hiding the tooltop that is still being hovered over
    if (this._hoveredElement === element) {
        this._tooltip.hide();
    }
};

arrays.Timeline.prototype.moveTooltip = function(left, top) {
    this._tooltip._container.css({ top, left });
};

/**
 * Set meta
 * @param {Object} meta
 * @public
 */
arrays.Timeline.prototype.setMeta = function(meta) {
    this._meta = arrays.clone(meta);
};

/**
 * Set included
 * @param {Object} included
 * @public
 */
arrays.Timeline.prototype.setIncluded = function(included) {
    this._included = arrays.clone(included);
};
