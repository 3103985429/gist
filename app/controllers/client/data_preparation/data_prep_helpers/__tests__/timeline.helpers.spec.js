import moment from 'moment';

import { processGroupedResults, getCountAggregationOperators } from '../timeline.helpers';
import { pokemon1, simpsons } from '../../../../../../internals/testing/backend/fixtures/datasets';
import { team1, team2 } from '../../../../../../internals/testing/backend/fixtures/teams';

const DATE_DURATION_DECADE = require('../../../../../config/dataTypes/date.config').DATE_DURATION_DECADE;

describe('timeline helpers', () => {
    describe('processGroupedResults', () => {
        it('should return grouped docs when sort by is date', () => {
            const options = {
                sortBy: 'original_air_date',
                defaultGroupByColumnName_humanReadable: 'original_air_date',
                groupBy_isDate: true,
                filterObj: {},
                groupSize: 10,
            };
            const dataSet = Object.assign({}, simpsons, { _team: team2 });
            const groupedResults = [
                {
                    startDate: moment.utc('1995-10-17T03:24:00').toDate(),
                    endDate: moment.utc('1995-10-21T03:24:00').toDate(),
                    total: 50,
                },
                {
                    startDate: moment.utc('1995-05-10T03:24:00').toDate(),
                    endDate: moment.utc('1995-05-10T03:24:00').toDate(),
                    total: 30,
                },
                {
                    startDate: moment.utc('1995-05-17T03:24:00').toDate(),
                    endDate: moment.utc('1995-06-21T03:24:00').toDate(),
                    total: 2,
                },
            ];
            const groupParams = { groupByUnit: 'months', filterDateFormat: 'YYYY-MM-DD' };
            const result = processGroupedResults(groupedResults, dataSet, options, groupParams);
            expect(result).toMatchSnapshot();
        });

        it('should return grouped docs when sort by is number', () => {
            const options = {
                sortBy: 'original_air_date',
                groupBy: 'Pokemon No_',
                sortByNumber: true,
                filterObj: {},
                groupSize: 10,
            };
            const dataSet = Object.assign({}, pokemon1, { _team: team1 });
            const groupedResults = [
                {
                    group: 100,
                    total: 50,
                },
                {
                    group: 200,
                    total: 30,
                },
                {
                    group: 500,
                    total: 2,
                },
            ];
            const groupParams = { groupNumber: 100 };
            const result = processGroupedResults(groupedResults, dataSet, options, groupParams);
            expect(result).toMatchSnapshot();
        });

        it('should return grouped docs when sort by is string', () => {
            const options = {
                groupBy: 'Type 1',
                filterObj: {},
                groupSize: 10,
            };
            const dataSet = Object.assign({}, pokemon1, { _team: team1 });
            const groupedResults = [
                {
                    group: 'Grass',
                    total: 50,
                },
                {
                    group: 'Water',
                    total: 30,
                },
                {
                    group: 'Fire',
                    total: 2,
                },
            ];
            const groupParams = {};
            const result = processGroupedResults(groupedResults, dataSet, options, groupParams);
            expect(result).toMatchSnapshot();
        });
    });

    describe('getCountAggregationOperators', () => {
        let aggregationOperators, groupParams;

        beforeEach(() => {
            aggregationOperators = [{ $match: { sample: { $exists: true } } }];
            groupParams = { groupIdQueryHelper: 'sample' };
        });

        it('should return default aggregation operators', () => {
            const options = {};
            const result = getCountAggregationOperators(aggregationOperators, groupParams, options);

            expect(result).toMatchSnapshot();
        });

        it('should return aggregation operators for empty group size', () => {
            const options = {
                groupSize: -1,
            };
            const result = getCountAggregationOperators(aggregationOperators, groupParams, options);

            expect(result).toMatchSnapshot();
        });

        it('should return aggregation operators for decade', () => {
            const options = {
                sortByDate: true,
                groupBy_realColumnName: DATE_DURATION_DECADE,
                sortBy_realColumnName: 'Name',
            };
            const result = getCountAggregationOperators(aggregationOperators, groupParams, options);

            expect(result).toMatchSnapshot();
        });
    });
});
