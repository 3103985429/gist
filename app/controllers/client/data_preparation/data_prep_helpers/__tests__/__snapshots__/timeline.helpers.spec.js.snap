// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`timeline helpers getCountAggregationOperators should return aggregation operators for decade 1`] = `
Array [
  Object {
    "$match": Object {
      "sample": Object {
        "$exists": true,
      },
    },
  },
  Object {
    "$unwind": "sample",
  },
  Object {
    "$group": Object {
      "_id": "sample",
    },
  },
]
`;

exports[`timeline helpers getCountAggregationOperators should return aggregation operators for empty group size 1`] = `
Array [
  Object {
    "$match": Object {
      "sample": Object {
        "$exists": true,
      },
    },
  },
  Object {
    "$unwind": "sample",
  },
  Object {
    "$group": Object {
      "_id": "sample",
      "count": Object {
        "$sum": 1,
      },
    },
  },
]
`;

exports[`timeline helpers getCountAggregationOperators should return default aggregation operators 1`] = `
Array [
  Object {
    "$match": Object {
      "sample": Object {
        "$exists": true,
      },
    },
  },
  Object {
    "$unwind": "sample",
  },
  Object {
    "$group": Object {
      "_id": "sample",
    },
  },
]
`;

exports[`timeline helpers processGroupedResults should return grouped docs when sort by is date 1`] = `
Array [
  Object {
    "endDate": 1995-10-21T03:24:00.000Z,
    "startDate": 1995-10-17T03:24:00.000Z,
    "total": 50,
    "viewAllFilter": Object {
      "original_air_date": "{\\"min\\":\\"1995-10-01\\",\\"max\\":\\"1995-10-31\\"}",
    },
    "viewAllUrl": "/simpsons/gallery?original_air_date=%7B%22min%22%3A%221995-10-01%22%2C%22max%22%3A%221995-10-31%22%7D",
  },
  Object {
    "endDate": 1995-05-10T03:24:00.000Z,
    "startDate": 1995-05-10T03:24:00.000Z,
    "total": 30,
    "viewAllFilter": Object {
      "original_air_date": "1995-05-10",
    },
    "viewAllUrl": "/simpsons/gallery?original_air_date=1995-05-10",
  },
  Object {
    "endDate": 1995-06-21T03:24:00.000Z,
    "startDate": 1995-05-17T03:24:00.000Z,
    "total": 2,
    "viewAllFilter": Object {
      "original_air_date": "{\\"min\\":\\"1995-05-01\\",\\"max\\":\\"1995-06-30\\"}",
    },
  },
]
`;

exports[`timeline helpers processGroupedResults should return grouped docs when sort by is number 1`] = `
Array [
  Object {
    "group": 100,
    "total": 50,
    "viewAllFilter": Object {
      "Pokemon No_": 100,
    },
    "viewAllUrl": "/pokemon/gallery?Pokemon%20No_=100",
  },
  Object {
    "group": 200,
    "total": 30,
    "viewAllFilter": Object {
      "Pokemon No_": 200,
    },
    "viewAllUrl": "/pokemon/gallery?Pokemon%20No_=200",
  },
  Object {
    "group": 500,
    "total": 2,
    "viewAllFilter": Object {
      "Pokemon No_": 500,
    },
  },
]
`;

exports[`timeline helpers processGroupedResults should return grouped docs when sort by is string 1`] = `
Array [
  Object {
    "group": "Grass",
    "total": 50,
    "viewAllFilter": Object {
      "Type 1": "Grass",
    },
    "viewAllUrl": "/pokemon/gallery?Type%201=Grass",
  },
  Object {
    "group": "Water",
    "total": 30,
    "viewAllFilter": Object {
      "Type 1": "Water",
    },
    "viewAllUrl": "/pokemon/gallery?Type%201=Water",
  },
  Object {
    "group": "Fire",
    "total": 2,
    "viewAllFilter": Object {
      "Type 1": "Fire",
    },
  },
]
`;
